### DISCUSSION TITLE
Dependency specifiers akin to PEP 508 from Python

### DISCUSSION THREAD

**Discussion by thernstig** [npm/rfcs#803] [RRFC]
## Problem

Often I want to install a package where I wished the package declared a subset of dependencies to be installed as well.

Python and [PEP 508](https://peps.python.org/pep-0508/) solves this. See also https://packaging.python.org/en/latest/specifications/dependency-specifiers/#dependency-specifiers.

An example of this can be seen [here](https://huggingface.co/docs/diffusers/installation):

```sh
pip install diffusers["torch"]
```

The `["torch"]` part tells us to install `diffusers` with a subset of tools, in this case also `torch`.

## A real-world use case
https://github.com/socketio/socket.io/ depends on the [`ws`](https://www.npmjs.com/package/ws) package. But the Node.js API now includes a [WebSocket implementation](https://nodejs.org/docs/v22.11.0/api/globals.html#websocket) of its own.

If I could then specify this:

```sh
npm install socket.io[no-ws]
```

I could skip that extra package and use the built-in one instead, without having to pollute my `node_modules/` with unnecessary packages.

**Is this a good or bad idea for an RFCS?**

**Comment by kylerneva** [comment]
npm already handles this through optionalDependencies, peerDependencies, and runtime feature detection, which allows packages to adapt to different environments without adding new install syntax or increasing dependency resolution complexity.

**Reply by thernstig** [reply]
@kylerneva Can you give an example of how to make this work for the `socket.io[no-ws]` example, that works equally simple for the end user as PEP 508?

I appreciate the answer, but it just mentions things where I have not found this to be possible. I.e. it gave no real answer to the exact way to do this.

**Reply by ljharb** [reply]
It's only up to the package author - and if they want to provide that, they'd have to make `ws` an optional peerDependency.

If they don't do that, you have no choice but to install `ws` (or use overrides to install something else that mimics the API)

**Reply by thernstig** [reply]
There is no "optional peerDependency" is there?

How do you achieve PEP 508 to work fine in a project? I cannot see how it is achievable right now.

**Reply by ljharb** [reply]
Yes, there is. You add the peer dep, and then in a parallel object "peerDependenciesMeta", you specify the dep name and the value as an object with "optional: true".

I have no idea because nobody outside Python ever knows or cares about PEPs. However, if I understand the OP correctly, you want (either, or both?) a package author/consumer to be able to get "not 100% of the production deps specified" installed. No, that's not possible - if you have such subsets, then your package is too big and should be broken up into smaller packages.

**Reply by thernstig** [reply]
@ljharb I think I explained it sufficiently in the original post. See `A real-world use case`.

> if you have such subsets, then your package is too big and should be broken up into smaller packages

That does not help in the example in the original post, or even if it did, it is not user friendly.

**Reply by ljharb** [reply]
That's right - you have to ask the maintainer to do it, or, you have to use another package.

**Reply by thernstig** [reply]
@ljharb if I understand you correct, your stance is that everything is fine as-is. My stance is that this would be good addition.

> I have no idea because nobody outside Python ever knows or cares about PEPs

You do. It is just named differently here. Here we call it [RFCs](https://github.com/npm/rfcs).
